File: qietr-indexer/poller/src/decode.ts
// =============================================================================
// decode.ts — TS mirror of qietr-indexer/geyser-plugin/src/decode.rs.
//
// The geyser plugin reads on-chain accounts off the validator's banking
// thread; this poller reads the same accounts over RPC instead. Both must
// decode byte-identically, so the borsh layouts and Anchor discriminators
// below are a 1:1 port of decode.rs. If a pool struct changes there, mirror
// it here — the discriminator self-tests in the geyser crate guard renames.
//
// Discriminator = sha256("account:<StructName>")[0..8].
// =============================================================================

import { createHash } from "node:crypto";

export const MERKLE_DEPTH = 20;
export const ROOT_HISTORY_LEN = 30;

function accountDisc(name: string): Buffer {
  return createHash("sha256").update(`account:${name}`).digest().subarray(0, 8);
}

function ixDisc(method: string): Buffer {
  return createHash("sha256").update(`global:${method}`).digest().subarray(0, 8);
}

export const DISC = {
  poolConfig: accountDisc("PoolConfig"),
  denomination: accountDisc("Denomination"),
  merkleTree: accountDisc("MerkleTree"),
  nullifierRecord: accountDisc("NullifierRecord"),
  depositIx: ixDisc("deposit"),
  withdrawIx: ixDisc("withdraw"),
} as const;

export type AccountKind =
  | "PoolConfig"
  | "Denomination"
  | "MerkleTree"
  | "NullifierRecord";

export function accountKind(data: Buffer): AccountKind | null {
  if (data.length < 8) return null;
  const head = data.subarray(0, 8);
  if (head.equals(DISC.poolConfig)) return "PoolConfig";
  if (head.equals(DISC.denomination)) return "Denomination";
  if (head.equals(DISC.merkleTree)) return "MerkleTree";
  if (head.equals(DISC.nullifierRecord)) return "NullifierRecord";
  return null;
}

// --- Borsh layouts (after the 8-byte discriminator) --------------------------

export interface Denomination {
  denomId: number;
  amountMicroUsdc: bigint;
  depositCount: bigint;
  vault: Buffer; // 32 raw bytes
}

export function decodeDenomination(data: Buffer): Denomination | null {
  // disc(8) | denom_id(1) | amount_micro_usdc(u64) | deposit_count(u64)
  //        | vault(32) | mint(32) | vault_bump(1) | bump(1)
  if (data.length < 8 + 1 + 8 + 8 + 32 + 32 + 1 + 1) return null;
  let o = 8;
  const denomId = data.readUInt8(o); o += 1;
  const amountMicroUsdc = data.readBigUInt64LE(o); o += 8;
  const depositCount = data.readBigUInt64LE(o); o += 8;
  const vault = Buffer.from(data.subarray(o, o + 32)); o += 32;
  return { denomId, amountMicroUsdc, depositCount, vault };
}

export interface MerkleTreeState {
  denomId: number;
  nextLeafIndex: bigint;
  rootCursor: number;
  latestRoot: Buffer | null; // 32 raw bytes, or null if cursor out of range
}

export function decodeMerkleTree(data: Buffer): MerkleTreeState | null {
  // disc(8) | denom_id(1) | next_leaf_index(u64)
  //        | filled_subtree([32;20]) | root_history([32;30])
  //        | root_cursor(1) | bump(1) | zero_hashes([32;20])
  const filledLen = 32 * MERKLE_DEPTH;
  const historyLen = 32 * ROOT_HISTORY_LEN;
  const minLen = 8 + 1 + 8 + filledLen + historyLen + 1 + 1 + 32 * MERKLE_DEPTH;
  if (data.length < minLen) return null;

  let o = 8;
  const denomId = data.readUInt8(o); o += 1;
  const nextLeafIndex = data.readBigUInt64LE(o); o += 8;
  o += filledLen; // skip filled_subtree
  const historyStart = o;
  o += historyLen;
  const rootCursor = data.readUInt8(o); o += 1;

  // latest_root() — bounds-checked exactly like decode.rs::latest_root.
  let latestRoot: Buffer | null = null;
  if (rootCursor < ROOT_HISTORY_LEN) {
    const start = historyStart + rootCursor * 32;
    latestRoot = Buffer.from(data.subarray(start, start + 32));
  }
  return { denomId, nextLeafIndex, rootCursor, latestRoot };
}

export interface NullifierRecord {
  denomId: number;
  nullifierHash: Buffer; // 32 raw bytes
  spentAtSlot: bigint;
}

export function decodeNullifierRecord(data: Buffer): NullifierRecord | null {
  // disc(8) | denom_id(1) | nullifier_hash(32) | spent_at_slot(u64) | bump(1)
  if (data.length < 8 + 1 + 32 + 8 + 1) return null;
  let o = 8;
  const denomId = data.readUInt8(o); o += 1;
  const nullifierHash = Buffer.from(data.subarray(o, o + 32)); o += 32;
  const spentAtSlot = data.readBigUInt64LE(o); o += 8;
  return { denomId, nullifierHash, spentAtSlot };
}

// --- deposit instruction --------------------------------------------------

export interface DepositIx {
  denomId: number;
  commitment: Buffer; // 32 raw bytes (big-endian on chain)
}

/**
 * Parse a `deposit` instruction's data, mirroring
 * geyser-plugin/src/lib.rs::handle_pool_ix.
 *
 * Wire format: discriminator(8) || denom_id(1) || commitment(32).
 * Returns null if the data isn't a deposit ix or is too short.
 */
export function parseDepositIx(data: Buffer): DepositIx | null {
  if (data.length < 8 + 1 + 32) return null;
  if (!data.subarray(0, 8).equals(DISC.depositIx)) return null;
  const denomId = data.readUInt8(8);
  const commitment = Buffer.from(data.subarray(9, 9 + 32));
  return { denomId, commitment };
}

// --- withdraw instruction -------------------------------------------------
//
// A withdraw ALSO appends a leaf to the tree: the change commitment
// (`public_signals[5]`), see qietr-pool/src/lib.rs::withdraw (tree.append).
// The indexer MUST capture it or the reconstructed tree diverges from chain
// and every later proof fails its root check.
//
// Wire format (Anchor, fixed-size arrays inline, no length prefixes):
//   disc(8) | denom_id(1) | nullifier_hash(32) | proof(256)
//          | public_signals(6 × 32)
// change_commitment = public_signals[5].
const WITHDRAW_PROOF_BYTES = 256;
const WITHDRAW_PUBLIC_SIGNAL_COUNT = 6;
const WITHDRAW_CHANGE_OFFSET = 8 + 1 + 32 + WITHDRAW_PROOF_BYTES + 5 * 32; // 457
const WITHDRAW_MIN_LEN = 8 + 1 + 32 + WITHDRAW_PROOF_BYTES + WITHDRAW_PUBLIC_SIGNAL_COUNT * 32; // 489

export interface WithdrawIx {
  denomId: number;
  changeCommitment: Buffer; // 32 raw bytes (big-endian on chain)
}

/**
 * Parse a `withdraw` instruction's change commitment. Returns null if the
 * data isn't a withdraw ix or is too short.
 */
export function parseWithdrawIx(data: Buffer): WithdrawIx | null {
  if (data.length < WITHDRAW_MIN_LEN) return null;
  if (!data.subarray(0, 8).equals(DISC.withdrawIx)) return null;
  const denomId = data.readUInt8(8);
  const changeCommitment = Buffer.from(
    data.subarray(WITHDRAW_CHANGE_OFFSET, WITHDRAW_CHANGE_OFFSET + 32),
  );
  return { denomId, changeCommitment };
}

File: qietr-indexer/geyser-plugin/src/decode.rs
// =============================================================================
// decode.rs — Anchor account discriminators + borsh layouts.
//
// Must stay byte-identical to qietr-pool/programs/qietr_pool/src/state.rs.
// If you add a field there, mirror it here. The discriminator is
// `sha256("account:<StructName>")[0..8]`; it only changes if the struct
// is renamed.
// =============================================================================

use borsh::BorshDeserialize;
use sha2::{Digest, Sha256};
use std::sync::OnceLock;

pub const MERKLE_DEPTH: usize = 20;
pub const ROOT_HISTORY_LEN: usize = 30;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AccountKind {
    PoolConfig,
    Denomination,
    MerkleTree,
    NullifierRecord,
}

fn anchor_disc(name: &str) -> [u8; 8] {
    let full = Sha256::digest(format!("account:{}", name).as_bytes());
    let mut out = [0u8; 8];
    out.copy_from_slice(&full[..8]);
    out
}

fn pool_config_disc() -> &'static [u8; 8] {
    static D: OnceLock<[u8; 8]> = OnceLock::new();
    D.get_or_init(|| anchor_disc("PoolConfig"))
}
fn denomination_disc() -> &'static [u8; 8] {
    static D: OnceLock<[u8; 8]> = OnceLock::new();
    D.get_or_init(|| anchor_disc("Denomination"))
}
fn merkle_tree_disc() -> &'static [u8; 8] {
    static D: OnceLock<[u8; 8]> = OnceLock::new();
    D.get_or_init(|| anchor_disc("MerkleTree"))
}
fn nullifier_record_disc() -> &'static [u8; 8] {
    static D: OnceLock<[u8; 8]> = OnceLock::new();
    D.get_or_init(|| anchor_disc("NullifierRecord"))
}

impl AccountKind {
    pub fn from_account_data(data: &[u8]) -> Option<Self> {
        if data.len() < 8 {
            return None;
        }
        let head = &data[..8];
        if head == pool_config_disc().as_slice() {
            return Some(Self::PoolConfig);
        }
        if head == denomination_disc().as_slice() {
            return Some(Self::Denomination);
        }
        if head == merkle_tree_disc().as_slice() {
            return Some(Self::MerkleTree);
        }
        if head == nullifier_record_disc().as_slice() {
            return Some(Self::NullifierRecord);
        }
        None
    }
}

#[derive(BorshDeserialize, Debug)]
pub struct Denomination {
    pub denom_id: u8,
    pub amount_micro_usdc: u64,
    pub deposit_count: u64,
    pub vault: [u8; 32],
    pub mint: [u8; 32],
    pub vault_bump: u8,
    pub bump: u8,
}

#[derive(BorshDeserialize, Debug)]
pub struct MerkleTree {
    pub denom_id: u8,
    pub next_leaf_index: u64,
    pub filled_subtree: [[u8; 32]; MERKLE_DEPTH],
    pub root_history: [[u8; 32]; ROOT_HISTORY_LEN],
    pub root_cursor: u8,
    pub bump: u8,
    pub zero_hashes: [[u8; 32]; MERKLE_DEPTH],
}

impl MerkleTree {
    /// Returns the active root (the latest one written via the ring-buffer),
    /// or `None` if `root_cursor` is out of range. The cursor is a `u8` read
    /// from on-chain data; a corrupt or malicious account could carry a value
    /// >= ROOT_HISTORY_LEN, and a direct index would panic on the validator's
    /// banking thread (taking the plugin — and potentially the validator —
    /// down). `get` makes this a safe, recoverable miss.
    pub fn latest_root(&self) -> Option<[u8; 32]> {
        self.root_history.get(self.root_cursor as usize).copied()
    }
}

#[derive(BorshDeserialize, Debug)]
pub struct NullifierRecord {
    pub denom_id: u8,
    pub nullifier_hash: [u8; 32],
    pub spent_at_slot: u64,
    pub bump: u8,
}

pub fn decode_denomination(data: &[u8]) -> Option<Denomination> {
    if data.len() < 8 {
        return None;
    }
    Denomination::try_from_slice(&data[8..]).ok()
}

pub fn decode_merkle_tree(data: &[u8]) -> Option<MerkleTree> {
    if data.len() < 8 {
        return None;
    }
    MerkleTree::try_from_slice(&data[8..]).ok()
}

pub fn decode_nullifier_record(data: &[u8]) -> Option<NullifierRecord> {
    if data.len() < 8 {
        return None;
    }
    NullifierRecord::try_from_slice(&data[8..]).ok()
}

/// Anchor instruction discriminator — `sha256("global:<method>")[0..8]`.
pub fn anchor_ix_disc(method: &str) -> [u8; 8] {
    let full = Sha256::digest(format!("global:{}", method).as_bytes());
    let mut out = [0u8; 8];
    out.copy_from_slice(&full[..8]);
    out
}

pub struct IxDescriptors {
    pub deposit: [u8; 8],
    pub withdraw: [u8; 8],
}

impl IxDescriptors {
    pub fn precomputed() -> Self {
        Self {
            deposit: anchor_ix_disc("deposit"),
            withdraw: anchor_ix_disc("withdraw"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // Reference values produced by `sha256("account:<Name>")[0..8]` and
    // `sha256("global:<method>")[0..8]` via Python:
    //   import hashlib
    //   hashlib.sha256(b"account:Denomination").digest()[:8].hex()
    //
    // Locking these in catches accidental renames of the pool's Anchor
    // structs — which would silently break ingest by skipping every
    // account write.
    #[test]
    fn account_discriminator_pool_config() {
        assert_eq!(
            *pool_config_disc(),
            [0x1a, 0x6c, 0x0e, 0x7b, 0x74, 0xe6, 0x81, 0x2b],
        );
    }

    #[test]
    fn account_discriminator_denomination() {
        assert_eq!(
            *denomination_disc(),
            [0xff, 0x5f, 0xf6, 0x28, 0x38, 0xa3, 0x6b, 0x55],
        );
    }

    #[test]
    fn account_discriminator_merkle_tree() {
        assert_eq!(
            *merkle_tree_disc(),
            [0x62, 0x33, 0x33, 0xe2, 0xa2, 0x14, 0x49, 0xd4],
        );
    }

    #[test]
    fn account_discriminator_nullifier_record() {
        assert_eq!(
            *nullifier_record_disc(),
            [0x38, 0x12, 0x39, 0xaf, 0x45, 0xca, 0xbd, 0x46],
        );
    }

    #[test]
    fn ix_discriminator_deposit() {
        // sha256("global:deposit")[:8] = f2 23 c6 89 52 e1 f2 b6
        assert_eq!(
            anchor_ix_disc("deposit"),
            [0xf2, 0x23, 0xc6, 0x89, 0x52, 0xe1, 0xf2, 0xb6],
        );
    }

    #[test]
    fn ix_discriminator_withdraw() {
        // sha256("global:withdraw")[:8] = b7 12 46 9c 94 6d a1 22
        assert_eq!(
            anchor_ix_disc("withdraw"),
            [0xb7, 0x12, 0x46, 0x9c, 0x94, 0x6d, 0xa1, 0x22],
        );
    }
}
